### PR TITLE
Shorten spin locks

### PR DESCRIPTION
**Pull request by avpotapov00**

### CODE REVIEW

**Review by ndkoval — CHANGES_REQUESTED** [review]
Thank you for the great work and well-documented code! I've started reviewing the PR, and have several minor concerns. Could you please address them before going further? After that, I will review the core algorithm.

Could you please also check how this change affects performance on both blocking (with spin-locks) and non-blocking code?

File: src/jvm/test/resources/expected_logs/obstruction_freedom_violation_events_cut.txt [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+|     compareAndSet(null,Node@3) at MSQueueBlocking.enqueue(ObtructionFreedomViolationTest.kt:28) |                                                                                                       |
+
+Detailed parallel part trace:
+|                                                                                                 | enqueue(-1)                                                                                           |
+|                                                                                                 |   enqueue(-1) at ObstructionFreedomViolationEventsCutTest.enqueue(SpinlockEventsCutTests.kt:25)       |
+|                                                                                                 |     <init>(-1,null,2,null) at MSQueueBlocking.enqueue(ObtructionFreedomViolationTest.kt:25)           |
+|                                                                                                 |       <init>(null) at MSQueueBlocking$Node.<init>(ObtructionFreedomViolationTest.kt:51)               |
```
[the remark this answers is not in the document]
Comment (avpotapov00) [reply src/jvm/test/resources/expected_logs/obstruction_freedom_violation_events_cut.txt:23]: 1. This bug caused not by my changes
(updating...)

File: src/jvm/test/org/jetbrains/kotlinx/lincheck/test/verifier/linearizability/ClocksTest.kt
Comment (avpotapov00) [inline src/jvm/test/org/jetbrains/kotlinx/lincheck/test/verifier/linearizability/ClocksTest.kt:1]: This test moved to SpinlockEventsCutTests.kt

**Comment by ndkoval** [comment]
Fixes #138

File: src/jvm/test/resources/expected_logs/spin_lock_in_incorrect_results_failure.txt [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+---
+Values in "[..]" brackets indicate the number of completed operations
+in each of the parallel threads seen at the beginning of the current operation
+---
+
+= The following interleaving leads to the error =
+Parallel part trace:
+|                      | c()                                                                                                 |
+|                      |   /* The following events repeat infinitely until thread switch */                                  |
```
Comment (ndkoval) [inline src/jvm/test/resources/expected_logs/spin_lock_in_incorrect_results_failure.txt:14]: Let's replace with "// The following events until the thread switch repeat infinitely"

File: src/jvm/test/resources/expected_logs/spin_lock_events_cut_single_action_cycle.txt [hunk 3]
```diff
@@ -0,0 +1,70 @@
+= The execution has hung, see the thread dump =
```
Comment (ndkoval) [inline src/jvm/test/resources/expected_logs/spin_lock_events_cut_single_action_cycle.txt:1]: There is no thread dump

File: src/jvm/test/resources/expected_logs/obstruction_freedom_violation_events_cut.txt [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+| enqueue(1)                                                                                              |                                                                                                        |
+|   enqueue(1) at ObstructionFreedomViolationEventsCutTest.enqueue(SpinlockEventsCutTests.kt:31)          |                                                                                                        |
+|     <init>(1,null,2,null) at MSQueueBlocking.enqueue(ObstructionFreedomViolationTest.kt:25)             |                                                                                                        |
+|       <init>(null) at MSQueueBlocking$Node.<init>(ObstructionFreedomViolationTest.kt:51)                |                                                                                                        |
+|     /* The following events repeat infinitely until thread switch */                                    |                                                                                                        |
```
Comment (ndkoval) [inline src/jvm/test/resources/expected_logs/obstruction_freedom_violation_events_cut.txt:32]: There is no thread switch below

File: src/jvm/test/resources/expected_logs/obstruction_freedom_violation_events_cut.txt [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+|     get(): Node@1 at MSQueueBlocking.enqueue(ObstructionFreedomViolationTest.kt:27)                     |                                                                                                        |
+|     compareAndSet(null,Node@3): false at MSQueueBlocking.enqueue(ObstructionFreedomViolationTest.kt:28) |                                                                                                        |
```
Comment (ndkoval) [inline src/jvm/test/resources/expected_logs/obstruction_freedom_violation_events_cut.txt:34]: We need to add a message indicating that an active lock has been detected

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/LincheckFailure.kt [hunk 6]
```diff
@@ -29,7 +29,7 @@ internal class IncorrectResultsFailure(
 
 internal class DeadlockWithDumpFailure(
     scenario: ExecutionScenario,
-    val threadDump: Map<Thread, Array<StackTraceElement>>,
+    val threadDump: Map<Thread, Array<StackTraceElement>>?,
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/LincheckFailure.kt:32]: Please leave a comment when the thread dump is `null` and when not

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/ParallelThreadsRunner.kt [hunk 7]
```diff
@@ -317,6 +317,7 @@ internal open class ParallelThreadsRunner(
 
     override fun close() {
         super.close()
+//        testThreadExecutions.forEach { it.cleanup() }
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/ParallelThreadsRunner.kt:320]: uncomment or remove

**Review by ndkoval — CHANGES_REQUESTED** [review]

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/FixedActiveThreadsExecutor.kt [hunk 8]
```diff
@@ -25,11 +25,6 @@ import java.util.concurrent.locks.*
  * possible, so that they should not be parked and unparked between invocations.
  */
 internal class FixedActiveThreadsExecutor(private val nThreads: Int, runnerHash: Int) : Closeable {
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/FixedActiveThreadsExecutor.kt:27]: Please make a separate PR with the executor fix

**Review by ndkoval — CHANGES_REQUESTED** [review]

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt [hunk 9]
```diff
@@ -707,8 +730,174 @@ abstract class ManagedStrategy(
...
     }
+
+    /**
+     * Detects loops, active locks and live locks when the same code location is visited too often.
+     * Its important that it's lifecycle is bounded by a certain scenario as it caches information about loops,
+     * based on the number of executions in threads and thread switches, which is reusable only in one scenario.
+     */
+    private inner class LoopDetector(private val hangingDetectionThreshold: Int) {
+        private var lastIThread = -1// no last thread
+        private val operationCounts = mutableMapOf<Int, Int>()
```
[the remark this answers is not in the document]
Comment (avpotapov00) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt]: Performance loss is insufficient, there is no need to do it right now

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/FixedActiveThreadsExecutor.kt (hunk 8 above)
Comment (avpotapov00) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/FixedActiveThreadsExecutor.kt:27]: Here it is: https://github.com/JetBrains/lincheck/pull/202/files

File: src/jvm/test/resources/expected_logs/obstruction_freedom_violation_events_cut.txt [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+|     <init>(1,null,2,null) at MSQueueBlocking.enqueue(ObstructionFreedomViolationTest.kt:25)             |                                                                                                        |
+|     // The following events repeat infinitely                                                           |                                                                                                        |
+|     get(): Node@1 at MSQueueBlocking.enqueue(ObstructionFreedomViolationTest.kt:27)                     |                                                                                                        |
+|     compareAndSet(null,Node@3): false at MSQueueBlocking.enqueue(ObstructionFreedomViolationTest.kt:28) |                                                                                                        |
+|     An active lock was detected                                                                         |                                                                                                        |
```
Comment (ndkoval) [inline src/jvm/test/resources/expected_logs/obstruction_freedom_violation_events_cut.txt:19]: Should we start all such messages with "//"?

File: src/jvm/test/resources/expected_logs/obstruction_freedom_violation_representation_with_no_cycle.txt [hunk 11]
```diff
@@ -0,0 +1,122 @@
+= The algorithm should be non-blocking, but an active lock is detected =
```
Comment (ndkoval) [inline src/jvm/test/resources/expected_logs/obstruction_freedom_violation_representation_with_no_cycle.txt:1]: You don't need "representation" in the filename

**Review by ndkoval — APPROVED** [review]

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/TracePoint.kt [hunk 12]
```diff
@@ -233,6 +244,10 @@ internal class CoroutineCancellationTracePoint(
     }
 }
 
+internal class SpinCycleStartTracePoint(iThread: Int, actorId: Int, callStackTrace: CallStackTrace): TracePoint(iThread, actorId, callStackTrace) {
+    override fun toStringImpl() =  "// The following events repeat infinitely"
```
Comment (zuevmaxim) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/TracePoint.kt:248]: Screenshot 2023-06-30 at 17 13 43

IDEA interprets `/` as a path separator
How about `[The following events repeat infinitely:]`?

**Comment by zuevmaxim** [comment]
The IDEA plugin becomes really slow after merging this branch

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/TracePoint.kt (hunk 12 above)
Comment (avpotapov00) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/TracePoint.kt:248]: I suggest to use well-known comment syntax: 
`// The following events repeat infinitely:`

**Review by zuevmaxim — APPROVED** [review]
